wasm/src/object.c

```diff
@@ -0,0 +1,40 @@
+#include "object.h"
+
+opa_value *opa_object_filter(opa_value *obj, opa_value *keys)
+{
+    if (opa_value_type(obj) != OPA_OBJECT)
+    {
+        return NULL;
+    }
+
+    if (opa_value_type(keys) != OPA_OBJECT && opa_value_type(keys) != OPA_ARRAY &&
+        opa_value_type(keys) != OPA_SET)
+    {
+        return NULL;
+    }
+
+    opa_object_t *r = opa_cast_object(opa_object());
+
+    for (opa_value *key = opa_value_iter(keys, NULL); key != NULL;
+         key = opa_value_iter(keys, key))
+    {
+        opa_value* k;
+        switch (opa_value_type(keys))
+        {
+        case OPA_OBJECT:
+        case OPA_SET:
+            k = key;
+            break;
+        case OPA_ARRAY:
+            k = opa_value_get(keys, key);
+        }
+
+        opa_object_elem_t *elem = opa_object_get(opa_cast_object(obj), k);
+        if (elem != NULL)
+        {
+            opa_object_insert(r, k, elem->v);
```

Comment:
If we ever start emitting free() calls in the policy wasm we'll need to implement reference counting or similar to handle cases like this.

Reply:
Right, that seems like the way to go.